Refactor InfoScanners

denizt opened the pull request:
We were not sharing any code between the scanners although they have some logic that can be easily shared. This patch is the initial refactoring to remedy the situation, however there is space left for more refactoring.

Montellese commented:
Please take a look at #6406 which will remove the `CThread` part from `CVideoInfoScanner` as it will be executed as an asynchronous job.

Jalle19 commented on xbmc/commons/InfoScanner.h:
```diff
@@ -0,0 +1,47 @@
+#include "filesystem/File.h"
+#include "threads/Thread.h"
+
+using namespace std;
```
Don't put this in headers

Memphiz commented:
thx @Jalle19 for seeing the evil before it happenend :)

denizt commented:
@Montellese that is a great step forward! :) Any reason not to make MusicInfoScanner, the code is similar between those too and yet different and this could be a good excuse to clean this up.

I will push a follow up change but should I rebase on that patch?

Jalle19 commented on xbmc/commons/InfoScanner.cpp:
```diff
@@ -0,0 +1,42 @@
+
+using namespace std;
+
+namespace XbmcCommons
```
I don't know what our conventions are, but if you add `using namespace XbmcCommons` you don't have to implement everything inside a namespace block, thus saving you one step of indentation.

Paxxi replied:
don't encourage using namespace, that just drags everything back into the global namespace.

Montellese replied:
@Paxxi: We are talking about the implementation of the class in that namespace here. IMO in that case a `using namespace` is OK. If you want to keep it this way at least change the indentation of the rest of the file to no base indentation.

Jalle19 commented on xbmc/commons/InfoScanner.h:
```diff
@@ -0,0 +1,47 @@
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
```
I am not absolutely sure but I think you need to have this directive at the first line in the file.

MartijnKaijser replied:
it doesn't really matter where it is as long as it's first before any actual code.

Paxxi replied:
Supposedly it's better for perf if it's first, some compilers will skip reading the file completely according to some talk from cppcon. I hardly think it matters though.

Montellese replied:
I think we have it at the very top in most header files (at least the ones I read frequently).

Paxxi replied:
Yup @Montellese we have it at the top in most but far from all and we even have it inside traditional ifdef guards :)

garbear replied:
putting #ifdef guards a the beginning of the header (theoretically) improves performance. `#pragma once` is processed differently and its location in the file won't affect performance

Montellese commented:
@denizt: I just wanted to know if it actually works and I'm more comfortable with the video library. And there we have "Mark as watched/unwatched" which doesn't exist in the music library so it was easier to test. Ideally the same would be done for the music library but it can/should be done in a separate PR.

If you are going to work on this I would do it on top of the commits in my PR because then you get the changes I made to the job manager etc.

Paxxi commented on xbmc/music/infoscanner/MusicInfoScanner.cpp:
```diff
@@ -42,7 +42,6 @@
 #include "settings/Settings.h"
 #include "FileItem.h"
 #include "guilib/LocalizeStrings.h"
-#include "utils/StringUtils.h"
```
why is this removed? StringUtils is still heavily used in the file.

denizt replied:
Util.h includes StringUtils, thx for noticing, I removed it from VideoInfoScanner too.

Montellese replied:
IMO you shouldn't rely on headers being included by other headers. If you explicitly use `StringUtils` then also include `utils/StringUtils.h`. That way someone else can remove all usage of `CUtil` and also the `Util.h` header without any additional problems.

denizt replied:
Good point. I am all for removing/minimising Util classes, with your approach it would be easier.

denizt replied:
@Montellese please take a look at #6518 , it removes some occurrences where this was the case

mkortstiege commented:
@denizt, just a quick fyi. https://github.com/xbmc/xbmc/pull/6406 has been merged and you can rebase on master now.

denizt commented:
I just pushed an updated patch (hopefully) addressing all the above comments.

Re: pragma once Interesting discussion and I actually look the closest class (Exception) as a base to match the code style
Re: namespaces, same as above

@Montellese I am not very familiar with the codebase so doing things take some time at this stage. I will try to take a stab at it though :)

denizt commented:
Rebased on master and fixed explicit include of StringUtils.

Anything else in this patch that needs further work on?

Montellese commented on xbmc/commons/InfoScanner.h:
```diff
@@ -0,0 +1,43 @@
+#include "utils/URIUtils.h"
+#include "filesystem/File.h"
+
+namespace XbmcCommons {
```
Please don't put this into `xbmc/commons` (or the `XbmcCommons` namespace) because that code is currently being built as a more or less separate library with very basic stuff in it (exceptions and logging) and the info scanner is nothing basic and very specific to Kodi.

denizt replied:
Ok, any suggestions which namespace I should add this to?

Montellese replied:
We don't really use namespaces that much so it doesn't have to be in any namespace. I'd move the files either into the `xbmc/` or the `xbmc/utils` directory.

Montellese commented on xbmc/commons/InfoScanner.h:
```diff
@@ -0,0 +1,43 @@
+namespace XbmcCommons {
+  class InfoScanner {
+  public:
+    virtual bool DoScan(const std::string& strDirectory) = 0;
```
Please re-order the methods to
* constructor (if present)
* destructor
* virtual methods (if present)
* normal methods (if present)

and add some empty lines between the method definitions for better readability.

Montellese commented on xbmc/commons/InfoScanner.h:
```diff
@@ -0,0 +1,43 @@
+     */
+    bool IsExcluded(const std::string& strDirectory, const std::vector<std::string> &regexps);
+  private:
+    bool HasNoMedia(const std::string& strDirectory) const
```
Please move the implementation to the CPP file. Then you can also save the two includes (but you are missing includes for `<string>` and `<vector>`).

denizt replied:
Will do, my intent here was to implicit-inline the method but that actually brings more complexity than its worth.

Montellese commented:
I'll provide the VS sync one the issues I've brought up have been taken care of.

denizt commented:
@Montellese I addressed all your comments and rebased on master.

Montellese commented on xbmc/InfoScanner.cpp:
```diff
@@ -0,0 +1,46 @@
+  return XFILE::CFile::Exists(noMediaFile);
+}
+
+bool InfoScanner::IsExcluded(const std::string& strDirectory, const std::vector<std::string> &regexps) {
```
The curly brace goes on a new line.

Montellese commented on xbmc/InfoScanner.cpp:
```diff
@@ -0,0 +1,46 @@
+
+#include "InfoScanner.h"
+#include "Util.h"
+#include "utils/log.h"
```
Please sort your includes in a hierarchically alphabetic order.

Montellese commented on xbmc/commons/Makefile:
```diff
@@ -1,6 +1,7 @@
 	Exception.cpp \
-	ilog.cpp
+	ilog.cpp \
+	InfoScanner.cpp
```
Wrong `Makefile`.

denizt commented:
@Montellese PTAL

Montellese commented on xbmc/InfoScanner.h:
```diff
@@ -0,0 +1,36 @@
+#include <string>
+#include <vector>
+
+class InfoScanner {
```
Curly brace on a new line (best make sure you do that for all of them).

denizt replied:
I missed this one:( To avoid more ping-pongs with the style nits I am trying to configure a formatter following http://kodi.wiki/view/Official:Code_guidelines_and_formatting_conventions

Any other conventions the wiki may be missing?

Montellese commented on xbmc/InfoScanner.cpp:
```diff
@@ -0,0 +1,47 @@
+ *
+ */
+
+#include "filesystem/File.h"
```
Please order them with the lowest hierarchy first and then by the sub-directories so
```cpp
#include "InfoScanner.h"
#include "URL.h"
#include "Util.h"
#include "filesystem/File.h"
#include "utils/log.h"
#include "utils/URIUtils.h"
```
(same goes for the includes in `xbmc/video/VideoInfoScanner.h`).

MartijnKaijser commented:
@denizt care to update this PR?

denizt commented:
@MartijnKaijser rebased and followed-up @Montellese 's comments

hudokkow commented:
jenkins build this please

denizt commented:
It fails to build on Windows and AppleTv, is there a quick way to add the new files to those platforms' project files?

mkortstiege commented:
appletv should be handling from within xcode. @Paxxi could you update the win32 project files?

Paxxi commented:
@mkortstiege I'll sort the VS projects, make sure to remove the hdhomerun stuff from xcode proj files though.

Paxxi commented:
cherry-pick Paxxi@805e79874f1a6bae12816615d5f2702f8e1d65a1 for VS projects

mkortstiege commented:
@Paxxi thanks :)

Paxxi commented:
just thought I'd mention it, InfoScanner doesn't really follow our naming rules, should be CInfoScanner but I'll leave it to you @mkortstiege to decide if it should be changed :)

mkortstiege commented:
Jep, `CInfoScanner` then for the sake of consistency.

denizt commented:
@mkortstiege Renamed to CInfoScanner and added the file to ATV Xcode. I merged @Paxxi's patch so the commit compiles.

mkortstiege commented:
jenkins build this please

denizt commented:
The ATV build still fails although I added the files to the project, or I thought I did... I thought we had stopped shipping Apple TV? Maybe @Memphiz can have a look?

MartijnKaijser commented:
Memphiz will be busy with RL for some time. @denizt can you start with a rebase to master? @mkortstiege guess we'll need your assistance ;) http://jenkins.kodi.tv/job/IOS-ATV2/4554/console

denizt commented:
@MartijnKaijser I rebased to master, no changes to Apple TV conf. :( I guess it will be taken off the build system some time soon anyway according to [this blog post](http://kodi.tv/farewell-to-future-appletv2-development/)?

MartijnKaijser commented:
jenkins build this please

MartijnKaijser commented:
@denizt seems the renaming of the files is gone

denizt commented:
@MartijnKaijser I didn't rename the files, I renamed the classes, see https://github.com/denizt/xbmc/commit/37b64861c59ce3a62f9322bc03999dfb4e843072#diff-961ad40f45701d05e6b1351abad54e90R28

Razzeee commented:
I would really love seeing this for Jarvis, can we get it up to speed again?

denizt commented:
If we are no longer building for ATV2, I can rebase and update this PR. The only problem left was building on that platform.

mkortstiege commented:
> If we are no longer building for ATV2, I can rebase and update this PR. The only problem left was building on that platform.

Jep, ATV2 was removed.

Razzeee commented:
@denizt can you rebase this?

denizt commented:
I wanted to rebase this but when I try to build the latest deps it fails at libsdl

```bash
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[3]: *** [build/SDLMain.lo] Error 1
make[2]: *** [macosx10.11_x86_64-target/build/.libs/libSDL.a] Error 2
make[1]: *** [libsdl] Error 2
```
Is this a known issue, any workarounds?

mkortstiege commented:
@denizt, sounds like dirty/broken depends.

denizt commented:
@mkortstiege I did a clean build and still no dice. Are the Jenkins instances doing a clean build every time?

Memphiz commented:
Jenkins builds as clean as we can get. I have no idea what sort of compile error you really have - the posted snipped is only part of the truth for sure because it could be a former error leading to this. Could you move your problem to the development forum and i can try to help you track it down ...

denizt commented:
So I had some time to look at this issue today, turns out it's 10.11 SDK, using 10.10 SDK I was able to compile deps without problems. It looks like its a common issue, as a workaround I used the 10.10 SDK from Xcode 6.4 as mentioned in https://github.com/js-platform/node-webrtc/issues/235#issuecomment-152447773

Memphiz commented:
mhh thats strange ... just built current master on osx 10.10 using xcode 7.2 and sdk 10.11 and what should i say:

Dependencies built successfully.

No issues ... will try your branch now ...

Memphiz commented:
scratch that - won't compile your branch as it doesn't even touch depends. All in all i see no issues with 10.11 sdk and keep watching the forum for other issues popping up eventually ...

denizt committed:
Refactor InfoScanners

We were not sharing any code between the scanners although they
have some logic that can be easily shared. This patch is the
initial refactoring to remedy the situation, however there is
space left for more refactoring.

denizt commented:
@Memphiz I will see if I can build on another El Capitan machine without problems next week

@mkortstiege I rebased, if all builds, should be good to go I guess?

MartijnKaijser commented:
Let's see. jenkins build this please

denizt commented:
@MartijnKaijser iOS build was failing as I didn't add the files to that target... I updated the commit and fixed that last issue

MartijnKaijser commented:
jenkins build this please

denizt commented:
@MartijnKaijser only Windows add ons build fails but I haven't touched it and it says non-blocking. I believe there it nothing left to do and the commit is good as as is? :smile:

Razzeee commented on xbmc/music/infoscanner/MusicInfoScanner.cpp:
```diff
@@ -439,8 +439,10 @@ bool CMusicInfoScanner::DoScan(const std::string& strDirectory)
   // Discard all excluded files defined by m_musicExcludeRegExps
-  std::vector<std::string> regexps = g_advancedSettings.m_audioExcludeFromScanRegExps;
-  if (CUtil::ExcludeFileOrFolder(strDirectory, regexps))
+
```
Unneeded line break

Razzeee commented on xbmc/music/infoscanner/MusicInfoScanner.h:
```diff
@@ -188,7 +189,7 @@ class CMusicInfoScanner : CThread, public IRunnable
   void GetAlbumArtwork(long id, const CAlbum &artist);
 
-  bool DoScan(const std::string& strDirectory);
+  virtual bool DoScan(const std::string& strDirectory);
```
I'm no expert on this, but should this really be virtual here? I would have thought we should loose the virtual and add an ```override```? But I may be out of my waters here.

Paxxi replied:
yup preferably it should be ```bool DoScan(const std::string& strDirectory) override;```
but it's not a blocker really

denizt replied:
Correct, let's be override-correct, push a new version rebased on master.

Razzeee commented:
jenkins build this please
